### PR TITLE
Fix bugs in ANN loss functions

### PR DESCRIPTION
**Pull request by hello-fri-end**
This PR aims to fix the bugs in ANN loss functions as pointed out in #2444.

- [x] L1 Loss.
- [x] Mean Bias Error.
- [x] Hinge Embedding Loss.
- [x] Huber Loss.
- [x] KL divergence.
- [x] Margin Ranking Loss.
- [x] Cosine Embedding Loss.
- [ ] Dice Loss : doens't make sense to add 'mean' reduction to this.
- [x] Earth Mover Distance.
- [x] Hinge Loss: correct implementation & reduction also already implemented.
- [x] Log cosh loss.
- [ ]  Mean Absolute Percentage Error: doens't make sense to add 'sum' reduction to this. Pytorch computes 'mean' & 'sqrt mean'.
- [x] Mean squared Error.
- [x] Soft Magin Loss: correct implementation & reduction also already implemented.
- [x] Multi Label SoftMargin Loss: correct implementation & reduction already implemented. 
- [x] Mean squared logarithmic Error.
- [x] NLL loss.
- [x] Poisson NLL loss.
- [x] Binary Cross Entropy Loss.
- [x] Reconstruction Loss.
- [x] Sigmoid Cross Entropy Loss.
- [ ] Triplet Margin Loss.

### CODE REVIEW

**Comment by hello-fri-end** [comment]
L1 Loss:
- [x] Replaced `mean` paramater with `reduction` that tells wheter the 'mean' or 'sum' will be returned.
- [x] Fixed the `forward` function of L1 Loss to take the absolute difference between `predictions` & `target`.
- [x] Updated the `backward` function to return a different output matrix, for different values of the `reduction` parameter.
- [x] Updated the tests.

**Comment by zoq** [comment]
Thanks for looking into this, is this ready for review it's marked as draft so maybe not.

**Comment by hello-fri-end** [comment]
> Thanks for looking into this, is this ready for review it's marked as draft so maybe not.

At the moment, only `L1 loss` is done. I'll add each loss function in a separate commit so that it's easier to review.

**Comment by hello-fri-end** [comment]
Mean Bias Error:

- [x] Added `reduction` parameter that tells wheter reduction 'sum' or 'mean' will be used.
- [x] Updated `forward` & `backward` function to return a different armadillo matrix based on differnet values of `reduction` parameter.
- [x] Added a non-trivial test.

**Comment by hello-fri-end** [comment]
Hinge Embedding Loss:

- [x] Added `reduction `parameter that tells wheter reduction 'sum' or 'mean' will be used.
- [x] Corrected the implementation of `Forward` & `Backward` pass to match with `torch.nn.HingeEmbeddingLoss`. 
- [x] Updated the tests.

**Comment by hello-fri-end** [comment]
Huber Loss: 

- [x]  Added `reduction` parameter that tells wheter reduction 'sum' or 'mean' will be used.
- [x] Updated `forward` & `backward` function to return a different armadillo matrix based on differnet values of `reduction` parameter.
- [x] Updated the test case to match with torch.nn.HuberLoss().

**Comment by hello-fri-end** [comment]
KL Divergence:

- [x] Added `reduction` parameter that tells wheter reduction 'sum' or 'mean' will be used.
- [x]  Corrected the implementation of `Forward` & `Backward` pass to match with `torch.nn.KLDivLoss()`.
- [x] Added a non-trivial test case.

**Comment by hello-fri-end** [comment]
Margin Ranking Loss:

- [x] Added `reduction` parameter that tells wheter reduction 'sum' or 'mean' will be used.
- [x] Corrected the `backward` method. Gradient was computed with respect to the labels while it should be calculated with respect to the two inputs `x1` & `x2`. 
- [x] Updated a non-trivial test that matches with `torch.nn.MarginRankingLoss()` .

**Comment by hello-fri-end** [comment]
Log CosH Error:

- [x]  Added `reduction` parameter that tells wheter reduction 'sum' or 'mean' will be used.
- [x]  Updated `forward` & `backward` function to return a different armadillo matrix based on differnet values of `reduction` parameter.
- [x] Added a test case for mean reduction.

**Comment by hello-fri-end** [comment]
@kartikdutt18 : I have a doubt regarding `Margin Ranking Loss` . It's defined as:

                         `loss(x1, x2, y) = max(0, −y ∗ ( x1 − x2 ) + margin)` 

In the current implementation, `x1` & `x2` are passed to the `forward` & `backward` function as one concatenated matrix, `predictions`, and `y` is the `target` matrix. Yesterday, I realized that the backward method is incorrect as it returns `df / dy` instead of `df / dx1` & `df / dx2` .So, I [updated](https://github.com/mlpack/mlpack/pull/3144/commits/46d491bc2f41f4ce640d0ca884af9e6b679ac0a9) the backward function to return the concatenated matrix `df / dx1` & `df / dx2`  (this is what the backward function of Pytorch returns) , but on a second thought this does not make much sense.  I was looking at `Cosine Embedding Loss` today & found that you've taken `y` as a hyperparameter, `x1` as the predictions matrix & `x2` as the target, so, you have only have to return `df / dx1` in the backward pass. I was wondering if it makes sense to do the same for `Hinge Embedding Loss`.  Let me know what you think about this.

cc: @zoq

**Comment by hello-fri-end** [comment]
> In the current implementation, `x1` & `x2` are passed to the `forward` & `backward` function as one concatenated matrix, `predictions`, and `y` is the `target` matrix. Yesterday, I realized that the backward method is incorrect as it returns `df / dy` instead of `df / dx1` & `df / dx2` .

The same problem exists in Triplet Margin Loss as well. It's defined as :

                                    `L(a, p, n ) = max ( d(a,p) - d(a,n) + margin, 0)`

Currently, `a` & `p` are passed to the `forward` & `backward` function as one concatenated matrix, `predictions` & `n` is the `target`. And the backward function only returns gradient with respect to `p`(not sure but I think it should atleast return grad w.r.t to `n` as well; Pytorch's backward returns grad w.r.t to all three).

File: src/mlpack/methods/ann/loss_functions/negative_log_likelihood.hpp [hunk 1]
```diff
@@ -19,7 +19,7 @@ namespace ann /** Artificial Neural Network. */ {
 
 /**
  * Implementation of the negative log likelihood layer. The negative log
- * likelihood layer expectes that the input contains log-probabilities for each
+ * likelihood layer expects that the input contains log-probabilities for each
  * class. The layer also expects a class index, in the range between 1 and the
  * number of classes, as target when calling the Forward function.
```
Comment (shubham1206agra) [inline src/mlpack/methods/ann/loss_functions/negative_log_likelihood.hpp:24]: Didn't we expect from 0 to num of class - 1?

**Comment by hello-fri-end** [comment]
@zoq @kartikdutt18 : This is ready for review. Everything is done except for `Hinge Embedding Loss` & `Triplet Margin Loss`.

**Comment by kartikdutt18** [comment]
> I was looking at Cosine Embedding Loss today & found that you've taken y as a hyperparameter, x1 as the predictions matrix & x2 as the target, so, you have only have to return df / dx1 in the backward pass. I was wondering if it makes sense to do the same for Margin Ranking Loss. Let me know what you think about this.

Maybe it makes more sense to return a concatenated matrix, I think Marcus would be the best person to answer this.

**Comment by zoq** [comment]
> > I was looking at Cosine Embedding Loss today & found that you've taken y as a hyperparameter, x1 as the predictions matrix & x2 as the target, so, you have only have to return df / dx1 in the backward pass. I was wondering if it makes sense to do the same for Margin Ranking Loss. Let me know what you think about this.
> 
> Maybe it makes more sense to return a concatenated matrix, I think Marcus would be the best person to answer this.

I have to take a closer look at the actual implementation, as I'm not sure I get the issue.

File: src/mlpack/methods/ann/loss_functions/binary_cross_entropy_loss.hpp [hunk 2]
```diff
@@ -77,9 +81,9 @@ class BCELoss
   //! Modify the epsilon.
   double& Eps() { return eps; }
 
-  //! Get the reduction.
+  //! Get the type of reduction used.
```
Comment (zoq) [inline src/mlpack/methods/ann/loss_functions/binary_cross_entropy_loss.hpp:84]: Perhaps we should use `Get the reduction type, represented as boolean (false 'mean' reduction, true 'sum' reduction).`?

File: src/mlpack/methods/ann/loss_functions/binary_cross_entropy_loss_impl.hpp [hunk 3]
```diff
@@ -60,6 +63,7 @@ void BCELoss<InputDataType, OutputDataType>::serialize(
     const uint32_t /* version */)
 {
   ar(CEREAL_NVP(eps));
+  ar(CEREAL_NVP(reduction));
```
Comment (zoq) [inline src/mlpack/methods/ann/loss_functions/binary_cross_entropy_loss_impl.hpp:66]: Nice catch!

File: src/mlpack/methods/ann/loss_functions/binary_cross_entropy_loss.hpp [hunk 4]
```diff
@@ -95,7 +100,8 @@ class BCELoss
   //! The minimum value used for computing logarithms and denominators
   double eps;
 
-  //! Reduction type. If true, performs mean of loss else sum.
+  //! Boolean value that tells if reduction 
+  //  is 'sum' or 'mean'.
```
Comment (zoq) [inline src/mlpack/methods/ann/loss_functions/binary_cross_entropy_loss.hpp:104]: Looks like this will fit on a single line.

File: src/mlpack/methods/ann/loss_functions/cosine_embedding_loss.hpp [hunk 5]
```diff
@@ -130,8 +133,9 @@ class CosineEmbeddingLoss
   //! Locally-stored value of similarity hyper-parameter.
   bool similarity;
 
-  //! Locally-stored value of takeMean hyper-parameter.
-  bool takeMean;
+//! Boolean value that tells if reduction
+  //  is 'sum' or 'mean'.
```
Comment (zoq) [inline src/mlpack/methods/ann/loss_functions/cosine_embedding_loss.hpp:137]: The indentation is off, also looks like this fits on a single line.

File: src/mlpack/methods/ann/loss_functions/binary_cross_entropy_loss_impl.hpp [hunk 6]
```diff
@@ -34,11 +34,13 @@ BCELoss<InputDataType, OutputDataType>::Forward(
...
-  ElemType loss = -arma::accu(target % arma::log(prediction + eps) +
+  ElemType lossSum = -arma::accu(target % arma::log(prediction + eps) +
       (1. - target) % arma::log(1. - prediction + eps));
+
   if (reduction)
-    loss /= prediction.n_elem;
-  return loss;
+    return lossSum;
+
+  return lossSum / target.n_elem;
```
Comment (zoq) [inline src/mlpack/methods/ann/loss_functions/binary_cross_entropy_loss_impl.hpp:43]: Shouldn't that be batch-size, same as in [cosine_embedding_loss_impl.hpp](https://github.com/mlpack/mlpack/pull/3144/files#diff-d525bfb306bbe31607580cff9c6d8721ff0a29959dcb4c895b449050b23bac32).

Comment (hello-fri-end) [reply src/mlpack/methods/ann/loss_functions/binary_cross_entropy_loss_impl.hpp:43]: In mean reduction, I'm dividing loss by the number of elements in output (Pytorch also does the same). The implementation of Cosine Embedding Loss in mlpack & Pytorch are slightly different. Pytorch takes `y` as target & in mlpack `y` is a hyperparamater. So, to match the results with Pytorch, the loss is divided by batchSize ( which is equivalent to dividing by no. of elements in y). However,  I'm not sure if the calculation of batchSize is correct. It's taken as `prediction.elem  / prediction.n_cols` which is equivalent to `prediction.n_rows * no. of slices` but don't we assume that each column represents a datapoint & not each row?

File: src/mlpack/methods/ann/loss_functions/log_cosh_loss_impl.hpp [hunk 7]
```diff
@@ -20,8 +20,9 @@ namespace mlpack {
 namespace ann /** Artificial Neural Network. */ {
 
 template<typename InputDataType, typename OutputDataType>
-LogCoshLoss<InputDataType, OutputDataType>::LogCoshLoss(const double a) :
-    a(a)
+LogCoshLoss<InputDataType, OutputDataType>::LogCoshLoss
+  (const double a, const bool reduction) :
```
Comment (zoq) [inline src/mlpack/methods/ann/loss_functions/log_cosh_loss_impl.hpp:24]:
```suggestion
LogCoshLoss<InputDataType, OutputDataType>::LogCoshLoss(
    const double a, const bool reduction) :
```

To match the style.

File: src/mlpack/methods/ann/loss_functions/log_cosh_loss_impl.hpp [hunk 8]
```diff
@@ -33,7 +34,13 @@ LogCoshLoss<InputDataType, OutputDataType>::Forward(
     const PredictionType& prediction,
     const TargetType& target)
 {
-  return arma::accu(arma::log(arma::cosh(a * (target - prediction)))) / a;
+  typename PredictionType::elem_type lossSum = 
+    arma::accu(arma::log(arma::cosh(a * (target - prediction)))) / a;
```
Comment (zoq) [inline src/mlpack/methods/ann/loss_functions/log_cosh_loss_impl.hpp:38]:
```suggestion
  typename PredictionType::elem_type lossSum = 
      arma::accu(arma::log(arma::cosh(a * (target - prediction)))) / a;
```

Needs to be tabbed twice see https://github.com/mlpack/mlpack/wiki/DesignGuidelines#line-length-and-wrapping for more details.

File: src/mlpack/methods/ann/loss_functions/mean_squared_error_impl.hpp [hunk 9]
```diff
@@ -41,16 +48,19 @@ void MeanSquaredError<InputDataType, OutputDataType>::Backward(
     const TargetType& target,
     LossType& loss)
 {
-  loss = 2 * (prediction - target) / target.n_cols;
+  loss = 2 * (prediction - target) ;
```
Comment (zoq) [inline src/mlpack/methods/ann/loss_functions/mean_squared_error_impl.hpp:51]:
```suggestion
  loss = 2 * (prediction - target);
```

Remove extra space.

File: src/mlpack/methods/ann/loss_functions/mean_squared_error_impl.hpp [hunk 10]
```diff
@@ -31,7 +32,13 @@ MeanSquaredError<InputDataType, OutputDataType>::Forward(
     const PredictionType& prediction,
     const TargetType& target)
 {
-  return arma::accu(arma::square(prediction - target)) / target.n_cols;
+  typename PredictionType::elem_type lossSum =
+    arma::accu(arma::square(prediction - target));
```
Comment (zoq) [inline src/mlpack/methods/ann/loss_functions/mean_squared_error_impl.hpp:36]:
```suggestion
  typename PredictionType::elem_type lossSum =
      arma::accu(arma::square(prediction - target));
```

Should be tabbed twice - https://github.com/mlpack/mlpack/wiki/DesignGuidelines#line-length-and-wrapping.

File: src/mlpack/methods/ann/loss_functions/mean_squared_logarithmic_error_impl.hpp [hunk 11]
```diff
@@ -32,8 +32,14 @@ MeanSquaredLogarithmicError<InputDataType, OutputDataType>::Forward(
     const PredictionType& prediction,
     const TargetType& target)
 {
-  return arma::accu(arma::square(arma::log(1. + target) -
-      arma::log(1. + prediction))) / target.n_cols;
+  typename PredictionType::elem_type lossSum =  
+    arma::accu(arma::square(arma::log(1. + target) -
+      arma::log(1. + prediction))) ;
```
Comment (zoq) [inline src/mlpack/methods/ann/loss_functions/mean_squared_logarithmic_error_impl.hpp:37]:
```suggestion
  typename PredictionType::elem_type lossSum =  
      arma::accu(arma::square(arma::log(1.0 + target) -
      arma::log(1.0 + prediction)));
```

File: src/mlpack/methods/ann/loss_functions/sigmoid_cross_entropy_error_impl.hpp [hunk 12]
```diff
@@ -42,7 +42,13 @@ SigmoidCrossEntropyError<InputDataType, OutputDataType>::Forward(
         std::log(1 + std::exp(-std::abs(prediction[i])));
   }
 
-  return maximum - arma::accu(prediction % target);
+  ElemType lossSum = 
+    maximum - arma::accu(prediction % target);
```
Comment (zoq) [inline src/mlpack/methods/ann/loss_functions/sigmoid_cross_entropy_error_impl.hpp:46]:
```suggestion
  ElemType lossSum = maximum - arma::accu(prediction % target);
```

File: src/mlpack/methods/ann/loss_functions/binary_cross_entropy_loss_impl.hpp (hunk 6 above)
Comment (zoq) [reply src/mlpack/methods/ann/loss_functions/binary_cross_entropy_loss_impl.hpp:43]: Yes, each column represents another sample.

**Comment by zoq** [comment]
What do you think should we merge this without updating the other three loss functions?

**Comment by zoq** [comment]
> > What do you think should we merge this without updating the other three loss functions?
> 
> Sounds good! We can merge this & I can open an issue for discussing the other three loss functions..

Sounds like a great idea to me.

**Review by zoq — APPROVED** [review]
Thanks for putting so much time into this.

**Review by kartikdutt18 — APPROVED** [review]

**Comment by zoq** [comment]
Thanks for the awesome contribution.

**Comment by rcurtin** [comment]
Sorry for coming across this issue late; great to have these issues fixed.  I did have one comment as I merge this into #2777 though--- @hello-fri-end would it have made more sense to instead create a class that could apply the mean reduction to any loss type?  So instead of adding a boolean parameter to every layer, you could just write a simple `MeanReduction<LossFunction>` class that computed `LossFunction`, then divided the loss function by the number of input columns?

**Comment by hello-fri-end** [comment]
> Sorry for coming across this issue late; great to have these issues fixed. I did have one comment as I merge this into #2777 though--- @hello-fri-end would it have made more sense to instead create a class that could apply the mean reduction to any loss type? So instead of adding a boolean parameter to every layer, you could just write a simple `MeanReduction<LossFunction>` class that computed `LossFunction`, then divided the loss function by the number of input columns?

There are some loss functions like `Mean Absolute Percentage Error` and `Dice Loss` for which mean reduction does not make any sense. How would we handle those? Also, this means our FFN instantiation will look like:

`FFN<MeanReduction<LossFunction>,InitializationRule<>> model;`

I guess, if the user doesn't want to apply mean reduction, they can still instantiate the network like: 

`FFN<LossFunction<>, InitializationRule> model; `

Hmm, seems like an interesting approach -- in future, we could add other reduction types as well. I guess, I'll wait till #2777 gets merged and then open a PR?
